tempesta_fw/apm.c

```diff
@@ -1151,14 +1496,318 @@ tfw_cfgop_apm_stats(TfwCfgSpec *cs, TfwCfgEntry *ce)
 	return 0;
 }
 
+static int
+tfw_cfgop_apm_server_failover(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	TfwApmHMCfg *hm_entry;
+	int code, limit, tframe;
+
+	if (tfw_cfg_check_val_n(ce, 3))
+		return -EINVAL;
+
+	if (ce->attr_n) {
+		TFW_ERR_NL("Unexpected attributes\n");
+		return -EINVAL;
+	}
+
+	if (tfw_cfgop_parse_http_status(ce->vals[0], &code)) {
+		TFW_ERR_NL("Unable to parse http code value: '%s'\n",
+			   ce->vals[0]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_parse_int(ce->vals[1], &limit)) {
+		TFW_ERR_NL("Unable to parse http limit value: '%s'\n",
+			   ce->vals[1]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_check_range(limit, 1, USHRT_MAX))
+		return -EINVAL;
+
+	if (tfw_cfg_parse_int(ce->vals[2], &tframe)) {
+		TFW_ERR_NL("Unable to parse http tframe value: '%s'\n",
+			   ce->vals[2]);
+		return -EINVAL;
+	}
+	if (tfw_cfg_check_range(tframe, 1, USHRT_MAX))
+		return -EINVAL;
+
+	hm_entry = kzalloc(sizeof(TfwApmHMCfg), GFP_KERNEL);
+	if (!hm_entry)
+		return -ENOMEM;
+
+	INIT_LIST_HEAD(&hm_entry->list);
+	hm_entry->code = code;
+	hm_entry->limit = limit;
+	hm_entry->tframe = tframe;
+	list_add(&hm_entry->list, &tfw_hm_codes_list);
+	++tfw_hm_codes_cnt;
+
+	return 0;
+}
+
+static void
+tfw_cfgop_apm_cleanup_server_failover(TfwCfgSpec *cs)
+{
+	TfwApmHMCfg *ent, *tmp;
+
+	list_for_each_entry_safe(ent, tmp, &tfw_hm_codes_list, list) {
+		list_del_init(&ent->list);
+		kfree(ent);
+	}
+	INIT_LIST_HEAD(&tfw_hm_codes_list);
+}
+
+static int
+tfw_cfgop_begin_apm_hm(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	TfwApmHM *hm;
+	size_t size;
+
+	if (tfw_cfg_check_val_n(ce, 1))
+		return -EINVAL;
+
+	if (ce->attr_n) {
+		TFW_ERR_NL("Unexpected attributes\n");
+		return -EINVAL;
+	}
+
+	list_for_each_entry(hm, &tfw_hm_list, list) {
+		if (!strcasecmp(hm->name, ce->vals[0])) {
+			TFW_ERR_NL("Duplicate health check entry: '%s'\n",
+				   ce->vals[0]);
+			return -EINVAL;
+		}
+	}
+
+	BUG_ON(tfw_hm_entry);
+	size = strlen(ce->vals[0]) + 1;
+	tfw_hm_entry = kzalloc(sizeof(TfwApmHM) + size, GFP_KERNEL);
+	if (!tfw_hm_entry) {
+		TFW_ERR_NL("Can't allocate health check entry '%s'\n",
+			   ce->vals[0]);
+		return -ENOMEM;
+	}
+	INIT_LIST_HEAD(&tfw_hm_entry->list);
+	tfw_hm_entry->name = (char *)(tfw_hm_entry + 1);
+	memcpy(tfw_hm_entry->name, ce->vals[0], size);
+	list_add(&tfw_hm_entry->list, &tfw_hm_list);
+
+	return 0;
+}
+
+static int
+tfw_cfgop_finish_apm_hm(TfwCfgSpec *cs)
+{
+	BUG_ON(!tfw_hm_entry);
+	BUG_ON(list_empty(&tfw_hm_list));
+	if (!tfw_hm_entry->codes && !tfw_hm_entry->crc32) {
+		TFW_ERR_NL("Response codes and crc32 values are not"
+			   " configured for '%s'\n", cs->name);
```

Comment:
The error message is confusing and doesn't report the actual reason for the error. Both of these configuration options are marked as `allow_none` which means that they can be skipped. So the actual error here, I believe, is that at least one of these (or both, for that matter) must be configured.

Reply:
@keshonok  Thank you for review.

Reply:
Corrected.